### PR TITLE
add real time feature when book tours

### PR DESCRIPTION
**Pull request by NT-NhatAnh**
## Related Tickets

- [#TicketID](https://edu-redmine.sun-asterisk.vn/issues/52235)

## WHAT (optional)

- create real time notification when users book a tour. This notification will send to all users currently online

## HOW

- create channel and let users join that channel. send data when users book a tour

## Evidence (Screenshot or Video)
image

### CODE REVIEW

**Comment by NT-NhatAnh** [comment]
ready

File: Gemfile [hunk 1]
```diff
@@ -21,6 +21,7 @@ gem "pagy"
 gem "puma", "~> 5.0"
 gem "rails", "~> 6.1.6"
 gem "rails-i18n"
+gem "redis"
```
Comment (manhnt-0960) [inline Gemfile]: redis để làm gì ?

File: app/controllers/tour_requests_controller.rb [hunk 2]
```diff
@@ -13,6 +13,7 @@ def create
     if tour_request.persisted?
       find_tour
       tour_request_success
+      broadcast_message current_user, tour_request
```
Comment (manhnt-0960) [inline app/controllers/tour_requests_controller.rb]: cho vào callback

File: app/javascript/channels/notifications_channel.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import consumer from "./consumer"
+
+consumer.subscriptions.create({ channel: "NotificationsChannel" },{
+  connected() {
+    console.log("connected")
```
Comment (manhnt-0960) [inline app/javascript/channels/notifications_channel.js]: bỏ log đi gái

File: app/javascript/channels/notifications_channel.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  },
+
+  disconnected() {
+    console.log("disconnected")
+  },
+
+  received(data) {
+    if (getCookieValue("user_id") != data.user_id) {
+      $("#popup_messages_name").html(data.user_name);
```
Comment (manhnt-0960) [inline app/javascript/channels/notifications_channel.js:12]: dùng `'` thôi e

Comment (manhnt-0960) [inline app/javascript/channels/notifications_channel.js:12]: dùng `'` thôi e

File: app/models/tour_request.rb [hunk 5]
```diff
@@ -26,12 +29,12 @@ class TourRequest < ApplicationRecord
   scope :lastest, ->{order(created_at: :desc)}
 
   def position
-    TourRequest.most_recent.index(self) + 1
+    User.find_by(id: user_id).tour_requests.most_recent.index(self) + 1
```
Comment (manhnt-0960) [inline app/models/tour_request.rb]: dùng luôn user cần gì find_by nữa

File: Gemfile (hunk 1 above)
Comment (NT-NhatAnh) [reply Gemfile]: em quên mất chưa xoá

**Comment by NT-NhatAnh** [comment]
ready

**Review by ghost — COMMENTED** [review]
mergred

**Pull request merged**